lib/Conversion/HWArithToHW/HWArithToHW.cpp

```diff
@@ -305,35 +306,63 @@ struct ArgResOpConversion : public OpConversionPattern<TOp> {
+      : OpConversionPattern<TOp>(tc, ctx), replacementFunc(f) {}
+
   LogicalResult
   matchAndRewrite(TOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
-    llvm::SmallVector<Type, 4> convResTypes;
+    llvm::SmallVector<Type> convResTypes;
     if (failed(this->getTypeConverter()->convertTypes(op->getResultTypes(),
                                                       convResTypes)))
       return failure();
-    // Use the generic builder to allow this pattern to apply to all ops.
-    rewriter.replaceOpWithNewOp<TOp>(op, convResTypes, adaptor.getOperands(),
-                                     op->getAttrs());
+    replacementFunc(rewriter, op, convResTypes, adaptor);
     return success();
   }
+
+  ReplacementFunc replacementFunc;
 };
 
-// Adds the ArgResOpConversion for 'TOp' to the set of conversion patterns, as
-// well as a legality check on the conversion status of the op's operands and
-// results.
-template <typename... TOp>
+// Adds the ArgResOpConversion for 'TOp' to the set of conversion patterns,
+// using TReplFunc as the replacement function inside the pattern, as well as a
+// legality check on the conversion status of the op's operands and results.
+template <typename TOp, typename TReplFunc>
 static void addOperandConversion(ConversionTarget &target,
                                  RewritePatternSet &patterns,
-                                 HWArithToHWTypeConverter &typeConverter) {
-  (patterns.add<ArgResOpConversion<TOp>>(typeConverter, patterns.getContext()),
-   ...);
-  target.addDynamicallyLegalOp<TOp...>([&](auto op) {
+                                 HWArithToHWTypeConverter &typeConverter,
+                                 TReplFunc replFunc) {
+  patterns.add<ArgResOpConversion<TOp>>(typeConverter, patterns.getContext(),
+                                        replFunc);
+
+  target.addDynamicallyLegalOp<TOp>([&](auto op) {
     return !typeConverter.hasSignednessSemantics(op->getOperandTypes()) &&
            !typeConverter.hasSignednessSemantics(op->getResultTypes());
   });
 }
 
+// Generic version of addOperandConversion (above) which uses the generic
+// operation builder signature.
+template <typename... TOp>
+static void addOperandConversion(ConversionTarget &target,
+                                 RewritePatternSet &patterns,
+                                 HWArithToHWTypeConverter &typeConverter) {
+  (addOperandConversion<TOp>(
+       target, patterns, typeConverter,
+       [](ConversionPatternRewriter &rewriter, TOp op,
+          llvm::ArrayRef<Type> convResTypes,
+          typename ArgResOpConversion<TOp>::OpAdaptor &adaptor) {
+         // Use the generic builder to allow this pattern to apply to all ops
+         // with default builders.
+         rewriter.replaceOpWithNewOp<TOp>(
+             op, convResTypes, adaptor.getOperands(), op->getAttrs());
+       }),
+   ...);
```

Comment:
WTF black c++ magic is this? What does it even do? I'm only familiar with the ellipses in function arguments.

Is it possible to rewrite this code in a way which doesn't use this syntax? For readability?

Reply:
Template fold expressions (https://en.cppreference.com/w/cpp/language/fold). It's actually found quite a few places in both LLVM and CIRCT (and the alternative looks much, much worse) - so I'd be inclined to keep it as-is.

Reply:
That was the link I couldn't find. Thanks!